scripts/shellcheck.sh

```diff
@@ -0,0 +1,253 @@
+                    if [[ -z "${EXCLUDE_CODES}" ]]; then
+                        EXCLUDE_CODES+="${CODE}"
+                    else
+                        EXCLUDE_CODES+=",${CODE}"
+                    fi
+                else
+                    # continue just don't save invalid error code
+                    printf "${RED}[Error]${RST_COLOR} Invalid error code entered: %s\n" "${optarg}"
+                    _err=1
+                fi
+                shift "$shft_cnt"
+                ;;
+            -v|--view)
+                VIEW="True"
+                shift "$shft_cnt"
+                ;;
+            -h|--help)
+                _help=1
+                ;;
+            *)
+                printf "${RED}[Error]${RST_COLOR} Unrecognized option: %s\n" "$1"
+                _err=1
+                shift "$shft_cnt"
+                ;;
+        esac
+
+        _invalid_arg_count=$(( "${_invalid_arg_count}" + 1 ))
+
+        # exit on help message
+        if [[ $_help -eq 1 ]]; then
+            show_usage
+            exit 0
+        fi
+        # continue on invalid arg, let user know but don't exit
+        if [[ "$_err" -eq 1 ]]; then
+            printf "${RED}[Error]${RST_COLOR} Invalid argument: %s\n" "${optarg}"
+        fi
+
+    done
+
+    if [[ "$_invalid_arg_count" -gt 0 ]]; then
+        # print the help message just once if one or more invalid arguments
+        echo; echo -e "${RED}[Error]${RST_COLOR} Some invalid arguments were detected."; echo
+        show_usage
+    fi
+}
+
+check_shellcheck_install() {
+    if ! command -v shellcheck &>/dev/null; then
+        echo -e "${RED}[ERROR]${RST_COLOR} shellcheck is not installed."
+        echo "Run 'sudo ./scripts/install-build-tools.sh' to install shellcheck."
+        exit 1
+    fi
+}
+
+get_num_cores() {
+    NUM_CORES=1
+    if [[ "$OSTYPE" == "linux-gnu"* ]]; then
+        NUM_CORES=$(grep -c ^processor /proc/cpuinfo)
+    elif [[ "$OSTYPE" == "darwin"* ]]; then
+        NUM_CORES=$(sysctl -n hw.ncpu)
+    else
+        NUM_CORES=1
+    fi
+}
+
+run_shellcheck() {
+    # check if git is installed
+    if command -v git &>/dev/null; then
+        echo "Using git ls-files to find shell scripts..."; echo
+        if [[ -z "$EXCLUDE_CODES" ]]; then
+            git ls-files '*.sh' | xargs -n 1 -P "$NUM_CORES" shellcheck -s bash > "$SHELLCHECK_REPORT"
+        else
+            git ls-files '*.sh' | xargs -n 1 -P "$NUM_CORES" shellcheck -s bash --exclude="$EXCLUDE_CODES" > "$SHELLCHECK_REPORT"
+        fi
+    else
+        echo "git is not installed. Using find to compile list of shell scripts..."; echo
+        if [[ -z "$EXCLUDE_CODES" ]]; then
+            find "$ROOT" -name '*.sh' -print0 | xargs -0 -n 1 -P "$NUM_CORES" shellcheck -s bash > "$SHELLCHECK_REPORT"
+        else
+            find "$ROOT" -name '*.sh' -print0 | xargs -0 -n 1 -P "$NUM_CORES" shellcheck -s bash --exclude="$EXCLUDE_CODES" > "$SHELLCHECK_REPORT"
+        fi
+    fi
+}
+
+check_report() {
```

Comment:
Can you clarify for me why you prefer having the report saved to a file instead of just sending to `stdout`? This adds a lot of complexity and requires you to deal with things like cache-invalidation.

Reply:
I allow for both. Send to file always, and `--view` option to see stdout.
When I am fixing errors I can search very easily in a file, but the stdout could exceed 1000 lines of output and be harder to view the errors of a code. Can grep the stdout but error messages take different amounts of line space so would have to play with `grep -B2 -A4` etc.
![shellcheck-file](https://github.com/user-attachments/assets/72437bc3-db62-4e9a-9aa9-947473b6654e)

Reply:
You barely need to add anything to support outputting to a file (because shells support redirection). Is overriding this default behavior (how almost all other command-line utilities work) of-use? Why not just redirect the output to a file as you need to?

Reply:
It creates two files now so this wouldn't be trivial anymore

Reply:
It now creates two files by running `shellcheck` twice every time, once at the requested severity level, and one bypassing the user's choice to make a "full" report. Why do this?

Reply:
The default for shellcheck is 'style' and that seems too strict to make a github action. My script defaults to 'error'.

There are over 100 messages of severity stricter than 'error' (warning, info, style) which I don't want myself or others to be blind to when turning in shell scripts or modifying them.

Very good to have the full report for commit 2 to mow down (warning, info, style) messages.

Reply:
Error is the least restrictive severity level as anything more strict gets ignored in pass/fail of the stage. 
What if we changed it to warning or info as default and used whichever one in the gh action? Can discuss in slack/zoom if need be.

Reply:
My phrasing might have been confusing. I meant less-restrictive in the sense of “print more diagnostics”. As I alluded to, I'm comfortable with “raising” the default from error to something higher, but I would ask that you make an argument for *why* we should prefer that. I appreciate unifying standards across the repository, and improving the CI/CD, *and* raising the bar for quality of code committed; but, for a research project predominantly focused on engineering novel solutions and not-at-all currently driven by productionization, these improvements are quality-of-life, not top-priority or of most research-interest. Adding friction for contributors absolutely *must* be justified; it isn't free.